src/Functions/array/arrayIndex.h

```diff
@@ -574,94 +624,247 @@ struct ArrayIndexGenericNullImpl
+                col_res->getData(),
+                null_map_data,
+                null_map_item);
+        else
+            return false;
+
+        block.getByPosition(result).column = std::move(col_res);
+        return true;
+    }
 
-        if (arguments.size() > 2)
+    /**
+     * 1. Obtain the right-side argument column @e C. If @e C is a non-const column (thus the argument is not constant),
+     * loop through all @e C's values.
+     * 2. Obtain the value's index.
+     * 3. Invoke the ArrayIndexNum*Impl to find the desired value
+     * 4. Fill the desired values in the resulting column
+     *
+     * Catches arguments of type T, LC(T), Nullable(LC(T)) and so on.
+     */
+    bool executeLowCardinality(Block & block, const ColumnNumbers & arguments, size_t result)
+    {
+        const ColumnArray * col_array = checkAndGetColumn<ColumnArray>(
+                block.getByPosition(arguments[0]).column.get());
+
+        if (!col_array)
+            return false;
+
+        /**
+         * Here we have four general cases:
+         * 1. LC(T), just search for T in the index.
+         * 2. Nullable(LC(T)), which is handled in nullMapsBuilder below. We can process this type as simple as the
+         * first one as all the *Impls take the null maps argument.
+         * 3. LC(Nullable(T)) and Nullable(LC(Nullable(T))). These cases are somewhat special as Nullable's getDataAt
+         * is slightly slower (due to nested column invocation).
+         *
+         * The array most outer nested type must be either LC(U) or Nullable(LC(U)).
+         * We do not care for LC(Nullable(U)) as it may be processed as V = Nullable(U) for LC(V).
+         */
+        const ColumnLowCardinality * col_array_nested_lc =
+            checkAndGetColumn<ColumnLowCardinality>(&col_array->getData());
+
+        if (!col_array_nested_lc)
         {
-            const auto & null_map1 = block.getByPosition(arguments[2]).column;
-            if (null_map1)
-                null_map_data = &assert_cast<const ColumnUInt8 &>(*null_map1).getData();
+            const ColumnNullable * const col_array_nested_nullable =
+                checkAndGetColumn<ColumnNullable>(&col_array->getData());
 
-            const auto & null_map2 = block.getByPosition(arguments[3]).column;
-            if (null_map2)
-                null_map_item = &assert_cast<const ColumnUInt8 &>(*null_map2).getData();
+            if (!col_array_nested_nullable)
+                return false;
+
+            col_array_nested_lc = checkAndGetColumn<ColumnLowCardinality>(
+                    &col_array_nested_nullable->getNestedColumn());
+
+            if (!col_array_nested_lc)
+                return false;
         }
 
-        const auto item_arg = block.getByPosition(arguments[1]).column.get();
+        auto col_res = ResultColumnType::create();
+        col_res->getData().resize_fill(col_array->getOffsets().size()); /// fill with default values
```

Comment:
IMO there is very little sense in pre-filling data with default (zero) values since we are going to overwrite those anyway in `ArrayIndexNumImpl::vectorCaseX`
Also that (with `ArrayIndexNumImpl`'s template parameter `ResizeRes`) add needless complexity.

Reply:
Ok, agree.